Add flags and config to run wails decoupled from project directory

bitterlox opened the pull request:
See issue #566

leaanthony reviewed (changes requested):
Thanks for this! I've added a couple of comments. What testing have you done? Platforms/Native vs Docker/etc?

bitterlox commented:
Changed behavior of -c as you suggested, also changed -o in the same way, and the second one doesn't need a check to verify the directory exists as the go tool will create it if it doesn't.
Fixed relative paths in config for frontend, moduleRoot and mainPackage. Fixed the same for mewn bundled assets, tho this change is a bit hacky, but anything more would need to be addressed in the mewn repo and it seems fine like this.
Also fixed a thing i didn't catch yesterday where serve would not find the built binary to run.
As for testing, i only tested on my machine, running linux. While making these changes i did avoid the docker parts of the code as i'm not sure how it should work with this.

bitterlox commented:
Found another small bug: i made all the paths absolute, but the functions to install the runtime joined the frontend path with projectDir, and it was causing the runtime to be mis-installed. Fixed and removed all the now unused projectDir params. Should be everything, tested both with my project and a fresh `wails init` one.  But please test on your end as well as i might have missed something.

leaanthony commented:
Will need to update the docker paths too to ensure cross-compiles don't fail.

leaanthony commented:
Hey @bitterlox - just checking in. Did you check out what's needed to get this working for cross-compilation via Docker?

bitterlox commented:
Hey, kind of got caught up in other stuff, i will check it out in the weekend.

bitterlox commented:
With these changes the cross compile should make use of the new functionality. However in testing it with a newly init'd project i was only able to cross compile to darwin, the linux/arm ones complete but do not generate an executable, and the windows one seems to hang.

leaanthony commented:
Yeah, that's not ideal! :-) Are you certain that it's the PR that's causing those hangs?

bitterlox commented:
No, i think it doesn't. I made a new branch off of 7e67562e19dbcbc29d554164ee31f1abe77c5b88 and re tried all the cross compiles, outcome looks the same. Also possibly windows works, it just takes a long time to pull all the stuff it needs.
So yeah i think everything works

leaanthony commented:
Sorry been slow to test this! Will have time this week 👍 Can you make a sample repo or instructions on how to make it fail currently? Thanks

bitterlox commented:
Hey, i just did a `wails init` to create a new project, and tried to cross compile it with the -x command, supplying the different platforms as an argument. To try it with the flags you can just navigate to the parent directory and pass in the `-c` flag with `./testProj/project.json` and the `-o` flag with `./testProj/build/testBin` or whichever flag values you want.

leaanthony commented:
Reopen

On Sat, 13 Feb 2021, 18:54 stale[bot], <notifications@github.com> wrote:

> This issue has been automatically marked as stale because it has not had
> recent activity. It will be closed if no further activity occurs. Thank you
> for your contributions.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/wailsapp/wails/pull/568#issuecomment-778579568>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO2SYAMVPI6GCNC7FB75ULS6YV5DANCNFSM4UFIO6CA>
> .
>

leaanthony commented:
Not forgotten about this. Will see if I can test today

leaanthony commented:
no

On Sat, 27 Mar 2021 at 00:49, stale[bot] ***@***.***> wrote:

> This issue has been automatically marked as stale because it has not had
> recent activity. It will be closed if no further activity occurs. Thank you
> for your contributions.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/wailsapp/wails/pull/568#issuecomment-808234694>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO2SYCOGNMMD7J267XXOJTTFSGIBANCNFSM4UFIO6CA>
> .
>

leaanthony commented on cmd/wails/6_serve.go:
```diff
@@ -10,16 +11,22 @@ import (
-		BoolFlag("f", "Force rebuild of application components", &forceRebuild)
+		BoolFlag("f", "Force rebuild of application components", &forceRebuild).
+		StringFlag("c", "Specify location of project.json", &configPath).
+		StringFlag("o", "Specify where the built executable should be placed", &outPath)
```
Is this asking for a directory or a /path/to/filename?

bitterlox replied:
`-o` is ultimately passed to `go build`, and the way that works is if you pass a directory it just places there the built executable, if you also pass a filename it also renames the executable. So can be either. `-c` should be a path, not dir

bitterlox commented on cmd/helpers.go:
```diff
@@ -239,7 +261,7 @@ func BuildNative(binaryName string, forceRebuild bool, buildMode string, project
 			}
 		}
-		buildCommand.Add("-o", filepath.Join("build", binaryName))
+		buildCommand.Add("-o", filepath.Join(outputDir, projectOptions.BinaryName))
```
it's been a while since i did these changes but i think yes i removed all relative paths otherwise it was breaking build.

bitterlox commented on cmd/helpers.go:
```diff
@@ -118,16 +134,22 @@ func InitializeCrossCompilation(verbose bool) error {
+	if filepath.IsAbs(buildDirectory) {
+		outDir = buildDirectory
+	} else {
+		outDir = filepath.Join(fs.Cwd(), buildDirectory)
```
this means, if `buildDirectory` is default, or set by the user, but in both cases absolute, we do nothing, otherwise, we get the absoulte path of it, so that it can be passed to docker and resolved correctly.

leaanthony commented:
Thanks for the clarifications. What platforms have you tested this on?

leaanthony commented:
I've just tested on Windows and it appears that if you try and build from outside of the module root, it complains. Is this an expected limitation? Also tested building up from a subdirectory and it appears we need to add updates to look for the build assets in the project root:
![image (7)](https://user-images.githubusercontent.com/1943904/113093984-a478b800-923c-11eb-9ac3-7ba7962b5cdd.png)

bitterlox commented:
> Thanks for the clarifications. What platforms have you tested this on?

linux and osx

> I've just tested on Windows and it appears that if you try and build from outside of the module root, it complains. Is this an expected limitation? Also tested building up from a subdirectory and it appears we need to add updates to look for the build assets in the project root:
> ![image (7)](https://user-images.githubusercontent.com/1943904/113093984-a478b800-923c-11eb-9ac3-7ba7962b5cdd.png)

did you specify module root and main pkg in project [options](https://github.com/bitterlox/wails/commit/a5a5a7469d9b1ccaac12ff60a6f673f25a017f43#diff-835f7544402d11b2f4f51edb1a86256f30b76571a199aa8ee61b28f15db1a405)?

leaanthony commented:
I did not, but isn't that what I'm passing in using `-c`? Perhaps if the module root isn't there it should save the relative path in the `project.json` file? Regarding `mainPackage`, what's an example of what that would be?

BTW: In xbar we solved this by just moving the whole wails app into a subdirectory and referenced local packages outside the go module root using `replace` in the application wide `go.mod`.

bitterlox commented:
Yeah perhaps it should save module root on wails init.
Mainpackage is essentially the path to the `package main` for the wails app, so that it knows where to go to build the application if it's located in a path not on the same level as the module root (for example under `cmd/`)
